File: dsaWithJava/tree/binarysearchtree/PairSum.java
package tree.binarysearchtree;

import java.util.ArrayList;

public class PairSum {
    public boolean isPairSum(Node root, int sum) {
        ArrayList<Integer> list = new ArrayList<>();
        list = isPairSum(root, list, sum);
        int head = 0, tail = list.size() - 1;
        while (head < tail) {
            if (list.get(head) + list.get(tail) == sum)
                return true;
            if (list.get(head) + list.get(tail) < sum)
                head++;
            else if (list.get(head) + list.get(tail) < sum)
                tail++;
        }
        return false;
    }

    private ArrayList<Integer> isPairSum(Node root, ArrayList<Integer> list, int sum) {
        if (root != null) {
            isPairSum(root.left, list, sum);
            list.add(root.data);
            isPairSum(root.right, list, sum);
        }
        return list;
    }
}

File: Python/tree/binary_search_tree/pair_sum.py
from node import Node


def findTarget(root: Node, k: int) -> bool:
    def __is_pair_sum(root: Node, lst: list, k: int) -> list:
        if root:
            __is_pair_sum(root.left, lst, k)
            lst.append(root.data)
            __is_pair_sum(root.right, lst, k)
        return lst

    lst = []
    lst = __is_pair_sum(root, lst, k)
    head, tail = 0, len(lst) - 1
    while head < tail:
        if lst[head] + lst[tail] == k:
            return True
        if lst[head] + lst[tail] < k:
            head += 1
        elif lst[head] + lst[tail] > k:
            tail += 1
    return False
